syncing/src/main/kotlin/maru/syncing/beaconchain/CLSyncServiceImpl.kt

```diff
@@ -61,45 +62,56 @@ class CLSyncServiceImpl(
     )
 
   override fun setSyncTarget(syncTarget: ULong) {
-    log.info("Syncing started syncTarget={}", syncTarget)
-    this.syncTarget.set(syncTarget)
+    check(started.get()) { "Sync service must be started before setting sync target" }
+
+    val oldTarget = this.syncTarget.getAndSet(syncTarget)
+    if (oldTarget != syncTarget) {
+      log.info("Syncing target updated from {} to {}", oldTarget, syncTarget)
+    }
 
     // If the pipeline is already running, we don't need to start a new one
-    if (pipeline == null) {
+    if (beaconChainPipeline.get() == null) {
       startSync()
     }
   }
 
+  @Synchronized
   private fun startSync() {
     val startBlock = beaconChain.getLatestBeaconState().latestBeaconBlockHeader.number + 1UL
     val pipeline = pipelineFactory.createPipeline(startBlock)
-    this.pipeline = pipeline
 
-    pipeline.start(executorService).handle { _, ex ->
-      if (ex != null && ex !is CancellationException) {
-        log.error("Sync pipeline failed, restarting", ex)
-        pipelineRestartCounter.increment()
-        this.pipeline = null
-        startSync()
-      } else {
-        log.info("Sync pipeline completed successfully")
-        this.pipeline = null
-        val completedSyncTarget = syncTarget.get() // Capture current target at completion
-        syncCompleteHanders.notifySubscribers(completedSyncTarget)
+    if (beaconChainPipeline.compareAndSet(null, pipeline)) {
+      pipeline.pipline.start(executorService).handle { _, ex ->
+        if (ex != null && ex !is CancellationException) {
+          log.error("Sync pipeline failed, restarting", ex)
+          pipelineRestartCounter.increment()
+          if (beaconChainPipeline.compareAndSet(pipeline, null)) {
+            startSync()
+          }
+        } else {
+          val completedSyncTarget = pipeline.target()
```

Comment:
My understanding is that this is the only reason why we introduced the `BeaconChainPipeline` data class. And since the target isn't really used, I'd suggest to rather remove the target from callback's signature and simplify the flow.